zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -129,15 +203,18 @@ impl Chain {
 trait UpdateWith<T> {
     /// Update `Chain` cumulative data members to add data that are derived from
     /// `T`
-    fn update_chain_state_with(&mut self, _: &T);
+    fn update_chain_state_with(&mut self, _: &T) -> Result<(), HistoryTreeError>;
```

Comment:
API usability: let's use a more general error type `ValidateContextError`
```suggestion
    fn update_chain_state_with(&mut self, _: &T) -> Result<(), ValidateContextError>;
```